src/main/java/org/genomicsdb/spark/sources/GenomicsDBInputPartitionReader.java

```diff
@@ -0,0 +1,356 @@
+import org.apache.spark.sql.types.StructField;
+import org.apache.spark.unsafe.types.UTF8String;
+import org.genomicsdb.reader.GenomicsDBFeatureReader;
+import org.genomicsdb.model.Coordinates;
+import org.genomicsdb.model.GenomicsDBExportConfiguration;
+import org.apache.spark.sql.types.StructType;
+
+import org.genomicsdb.spark.GenomicsDBInput;
+import org.genomicsdb.spark.GenomicsDBVidSchema;
+
+import org.json.simple.parser.ParseException;
+import scala.collection.JavaConverters;
+
+import java.io.IOException;
+import java.lang.reflect.Array;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * Used by an executor to read data from storage, as an internal row based on the schema. 
+ **/
+public class GenomicsDBInputPartitionReader implements PartitionReader<InternalRow> {
+
+  private GenomicsDBFeatureReader<VariantContext, PositionalBufferedStream> fReader;
+  private CloseableTribbleIterator<VariantContext> iterator;
+  private GenomicsDBInputPartition inputPartition;
+
+  public GenomicsDBInputPartitionReader(GenomicsDBInputPartition iPartition) {
+    inputPartition = iPartition;
+    String loader = inputPartition.getLoader();
+    String query = inputPartition.getQuery();
+    GenomicsDBExportConfiguration.ExportConfiguration exportConfiguration;
+    try {
+      exportConfiguration =
+          GenomicsDBInput.createTargetExportConfigurationPB(
+              query, inputPartition.getPartitionInfo(), 
+              inputPartition.getQueryInfoList(), inputPartition.getQueryIsPB());
+    } catch (ParseException | IOException e) {
+      e.printStackTrace();
+      exportConfiguration = null;
+    }
+
+    try {
+      fReader =
+          new GenomicsDBFeatureReader<>(
+              exportConfiguration,
+              (FeatureCodec<VariantContext, PositionalBufferedStream>) new BCF2Codec(),
+              Optional.of(loader));
+      this.iterator = fReader.iterator();
+    } catch (IOException e) {
+      e.printStackTrace();
+      fReader = null;
+    }
+  }
+
+  public InternalRow get() {
+    VariantContext val;
+    val = (VariantContext) this.iterator.next();
+
+    // get alt allele info
+    List<Allele> vcAltAlleles = val.getAlternateAlleles();
+    UTF8String[] altAlleles = null;
+    if (vcAltAlleles != null && vcAltAlleles.size() > 0) {
+      altAlleles = new UTF8String[vcAltAlleles.size()];
+      for (int i = 0; i < vcAltAlleles.size(); i++) {
+        altAlleles[i] = UTF8String.fromString(vcAltAlleles.get(i).getBaseString());
+      }
+    }
+
+    UTF8String[] sampleNames = new UTF8String[val.getNSamples()];
+    UTF8String[] genotypes = new UTF8String[val.getNSamples()];
+
+    // get sample name and genotype info
+    int i = 0;
+    for (Genotype g: val.getGenotypesOrderedByName()) {
+      sampleNames[i] = UTF8String.fromString(g.getSampleName());
+      genotypes[i++] = UTF8String.fromString(g.getGenotypeString());
+    }
+
```

Comment:
Mental note - it might be better to produce data in the format required by DataFrame from the C++ side rather than a conversion from VariantContext objects.

Reply:
Agreed, I think this is on the stack for release 0.1.2